### ISSUE TITLE
Regex causing exponential runtime+mem usage

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10323]
Migrated from [rt.perl.org#74484](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=74484) (status was 'resolved')

Searchable as RT74484$

**Comment by p5pRT** [comment]
### From @apocalypse

#### Created by cpan@ubuntu-server64.0ne.us



Hello\,

&nbsp;    I've been smoking CPAN modules for CPANTesters and noticed that my 
bots got OOMKilled
from time to time\. At first I thought it was the bot itself and blamed 
my VM system for
being wonky\. After it happened one too many times I gave up and decided 
to do a full\-blown
investigation\. Turned out the culprit was Router&#8203;::Generic's testsuite\. 
Whenever it runs\,
the parent POE&#8203;::Component&#8203;::SmokeBox process kept dying\.

&nbsp;    The problem is that SmokeBox uses String&#8203;::Perl&#8203;::Warnings to check 
incoming strings for sanity\.
S&#8203;::P&#8203;::W uses Regexp&#8203;::Assemble to build a super\-big regex of \*ALL\* Perl 
warnings and uses that to match
against the string\.

&nbsp;    I'm using Regexp&#8203;::Assemble v0\.34 and String&#8203;::Perl&#8203;::Warnings v1\.04 in 
the app\. After eliminating
the POE\+CPAN stuff\, I've come up with a testcase that is guaranteed to 
consume all RAM on my machine
and take a long time to complete\. I'm not a regex whiz so I have no idea 
why it's happening so I've
decided to post this perlbug so the right people can look at it\.

&nbsp;    After playing around with this script I made some observations&#8203;:

&nbsp;    \* I've tested that this blows up on 5\.12\.0 and 5\.10\.0\.
&nbsp;    \* Leaving the string as\-is will result in the pathological behavior 
in the regex engine
&nbsp;    \* Changing the string to "a" results in constant memory usage but 
slightly increased
&nbsp;        runtime each loop \( this makes sense\, as the string is getting 
bigger and bigger \)
&nbsp;    \* Having spaces in the string seems to be the culprit\, as doing "a 
a" results in mem usage going up
&nbsp;        but not as fast as the default\!
&nbsp;    \* At first I thought it was the \\r but having a string of "\\ra" 
behaves the same as "a"

&nbsp;        string "a"&#8203;:    Loop 300&#8203;: 0\.00176382064819336s rss\(9340 kB\)
&nbsp;        string "a a"&#8203;:    Loop 300&#8203;: 0\.0624589920043945s rss\(23836 kB\)
&nbsp;        string default&#8203;:    Loop 300&#8203;: 1\.41920399665833s rss\(238724 kB\)

&nbsp;    \* I thought it might be my script\, so I bypassed the loop and just 
did "foreach my $c \( 1000 \)" and it
&nbsp;        promptly got OOMKilled\. Trying it with 300 resulted in pretty 
much the same runtime/mem usage
&nbsp;        as the one above for default\.

&nbsp;    \* Look at the DATA section of the testcase\, and wonder why it 
suddenly gets exponential after around
&nbsp;        520 loops for runtime\.\.\. However\, the memory increase is pretty 
constant per loop\.\.\. Also\, it
&nbsp;        somehow "stabilizes" after 520 loops \- did the perlinternals 
know memory is running out and
&nbsp;        behaved differently?

&nbsp;    Hope you'll have as much fun playing with this testcase as I did :\)

\#\!/usr/bin/perl
use strict; use warnings;

use String&#8203;::Perl&#8203;::Warnings qw\( is\_warning \);
use Time&#8203;::HiRes qw\( time \);

\# Let Regexp&#8203;::Assemble init the regex
my $result = is\_warning\( '' \);

\# Loop over the test string and find out where Perl hits the OOMKill 
problem \( my VM has 768M of RAM \)
foreach my $c \( 0 \.\. 1000 \) \{
&nbsp;    \# Construct the test string
&nbsp;    my $string = '';
&nbsp;    for \( 0 \.\. $c \) \{
&nbsp;        $string \.= "\\radd\_route&#8203;: $\_/1000";
&nbsp;    \}

&nbsp;    \# Time the regex engine
&nbsp;    my $starttime = time;
&nbsp;    $result = is\_warning\( $string \);
&nbsp;    my $stoptime = time;

&nbsp;    \# Print diag messages every 10 loops
&nbsp;    unless \( $c % 10 \) \{
&nbsp;        \# Argh\, get the mem usage
&nbsp;        my $rss = get\_memory\(\);
&nbsp;        warn "Loop $c&#8203;: " \. \($stoptime \- $starttime\) \. "s rss\($rss\)\\n";
&nbsp;    \}
\}

sub get\_memory \{
&nbsp;    \# hackish way\, but works for now :\(
&nbsp;    my @&#8203;data = \`cat /proc/$$/status\`;
&nbsp;    foreach my $line \( @&#8203;data \) \{
&nbsp;        if \( $line =~ /^VmRSS&#8203;:\\s\+\(\.\+\)$/ \) \{
&nbsp;            return $1;
&nbsp;        \}
&nbsp;    \}
\}

1;

\_\_DATA\_\_

I've ran this program on my VM and here's the output&#8203;:
\[cpan@&#8203;ubuntu\-server64 ~\]$ perls/perl\_5\.12\.0\_default/bin/perl regex\_crash\.pl
Loop 0&#8203;: 0\.0791380405426025s rss\(9336 kB\)
Loop 10&#8203;: 0\.00212907791137695s rss\(9340 kB\)
Loop 20&#8203;: 0\.00815701484680176s rss\(9340 kB\)
Loop 30&#8203;: 0\.015045166015625s rss\(9800 kB\)
Loop 40&#8203;: 0\.0259301662445068s rss\(11144 kB\)
Loop 50&#8203;: 0\.0399010181427002s rss\(13280 kB\)
Loop 60&#8203;: 0\.0568771362304688s rss\(16080 kB\)
Loop 70&#8203;: 0\.0785748958587646s rss\(19384 kB\)
Loop 80&#8203;: 0\.096937894821167s rss\(23176 kB\)
Loop 90&#8203;: 0\.130855083465576s rss\(27480 kB\)
Loop 100&#8203;: 0\.153497934341431s rss\(32284 kB\)
Loop 110&#8203;: 0\.180907964706421s rss\(37600 kB\)
Loop 120&#8203;: 0\.216251134872437s rss\(43448 kB\)
Loop 130&#8203;: 0\.258400201797485s rss\(49816 kB\)
Loop 140&#8203;: 0\.313691854476929s rss\(56716 kB\)
Loop 150&#8203;: 0\.344374895095825s rss\(64144 kB\)
Loop 160&#8203;: 0\.387225151062012s rss\(72096 kB\)
Loop 170&#8203;: 0\.438238859176636s rss\(80572 kB\)
Loop 180&#8203;: 0\.498692989349365s rss\(89580 kB\)
Loop 190&#8203;: 0\.550199031829834s rss\(99112 kB\)
Loop 200&#8203;: 0\.604217052459717s rss\(109168 kB\)
Loop 210&#8203;: 0\.675564050674438s rss\(119756 kB\)
Loop 220&#8203;: 0\.791208028793335s rss\(130864 kB\)
Loop 230&#8203;: 0\.850717067718506s rss\(142504 kB\)
Loop 240&#8203;: 0\.894576072692871s rss\(154668 kB\)
Loop 250&#8203;: 0\.97498893737793s rss\(167360 kB\)
Loop 260&#8203;: 1\.04289603233337s rss\(180584 kB\)
Loop 270&#8203;: 1\.14439296722412s rss\(194328 kB\)
Loop 280&#8203;: 1\.20763897895813s rss\(208596 kB\)
Loop 290&#8203;: 1\.31933212280273s rss\(223396 kB\)
Loop 300&#8203;: 1\.41920399665833s rss\(238724 kB\)
Loop 310&#8203;: 1\.48294711112976s rss\(254572 kB\)
Loop 320&#8203;: 1\.59841012954712s rss\(270952 kB\)
Loop 330&#8203;: 1\.70976114273071s rss\(287856 kB\)
Loop 340&#8203;: 1\.80522298812866s rss\(305284 kB\)
Loop 350&#8203;: 1\.92089986801147s rss\(323248 kB\)
Loop 360&#8203;: 2\.01572108268738s rss\(341728 kB\)
Loop 370&#8203;: 2\.15599703788757s rss\(360748 kB\)
Loop 380&#8203;: 2\.27072596549988s rss\(380276 kB\)
Loop 390&#8203;: 2\.40587496757507s rss\(400348 kB\)
Loop 400&#8203;: 2\.51369595527649s rss\(420940 kB\)
Loop 410&#8203;: 2\.63296818733215s rss\(442056 kB\)
Loop 420&#8203;: 2\.76079702377319s rss\(463700 kB\)
Loop 430&#8203;: 2\.90137791633606s rss\(485872 kB\)
Loop 440&#8203;: 3\.00859093666077s rss\(508572 kB\)
Loop 450&#8203;: 3\.19537901878357s rss\(531792 kB\)
Loop 460&#8203;: 3\.36626696586609s rss\(555548 kB\)
Loop 470&#8203;: 3\.49290084838867s rss\(579828 kB\)
Loop 480&#8203;: 3\.5871410369873s rss\(604636 kB\)
Loop 490&#8203;: 3\.73700594902039s rss\(629964 kB\)
Loop 500&#8203;: 3\.89036107063293s rss\(655824 kB\)
Loop 510&#8203;: 4\.06298685073853s rss\(682212 kB\)
Loop 520&#8203;: 4\.23448181152344s rss\(709124 kB\)
Loop 530&#8203;: 6\.80471897125244s rss\(705788 kB\)
Loop 540&#8203;: 10\.8200550079346s rss\(709936 kB\)
Loop 550&#8203;: 14\.4348080158234s rss\(698272 kB\)
Loop 560&#8203;: 21\.236438035965s rss\(700008 kB\)
Loop 570&#8203;: 25\.1975688934326s rss\(700464 kB\)
Killed

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.12.0:

Configured by cpan at Mon Apr 12 10:44:17 MST 2010.

Summary of my perl5 (revision 5 version 12 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=2.6.31-14-server, archname=x86_64-linux
    uname='linux ubuntu-server64 2.6.31-14-server #48-ubuntu smp fri oct 
16 15:07:34 utc 2009 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/cpan/perls/perl_5.12.0_default'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include'
    ccversion='', gccversion='4.4.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.10.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.10.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib 
-fstack-protector'

Locally applied patches:
    


@INC for perl 5.12.0:
    /home/cpan/perls/perl_5.12.0_default/lib/site_perl/5.12.0/x86_64-linux
    /home/cpan/perls/perl_5.12.0_default/lib/site_perl/5.12.0
    /home/cpan/perls/perl_5.12.0_default/lib/5.12.0/x86_64-linux
    /home/cpan/perls/perl_5.12.0_default/lib/5.12.0
    .


Environment for perl 5.12.0:
    HOME=/home/cpan
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Apr 19\, 2010 at 01&#8203;:34&#8203;:02AM \-0700\, perl@&#8203;0ne\.us wrote&#8203;:
> I've been smoking CPAN modules for CPANTesters and noticed that my bots
> got OOMKilled from time to time\.

Thanks for the report\.

It appears there are two \(mostly\) unrelated issues here\. First is the
performance\, which is quadratic in string length\.

This is due to String&#8203;::Perl&#8203;::Warnings creating a regexp that looks
essentially like /\.\*?\.\.\.\./\. This is equivalent to /^\.\*?\.\*?\.\.\.\.\./
\(note the anchor\) and on repeated backtracking the two \.\*? make it go
quadratic\. The fix for the String&#8203;::Perl&#8203;::Warnings module's author is to
either remove the initial \.\*?\, or add an initial anchor\.

The second issue is the memory usage\, which again appears to be quadratic
on string length\. This is due to the trie code leaking SVs on the tmps
stack \(which then accumulate quadratically with the quadratic backtracking
caused by the first issue\)\. I've currently got some ideas on how to rework
that bit of code to not need to create temp SVs to store the accept
states\, but in the meantime I've added a TODO test with commit
8bd05d90e59a554e79d73edf32bf35547dbc2c40


\-\- 
I don't want to achieve immortality through my work\.\.\. I want to achieve
it through not dying\.
&nbsp;    \-\- Woody Allen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Apr 21\, 2010 at 12&#8203;:24&#8203;:21PM \+0100\, Dave Mitchell wrote&#8203;:
> On Mon\, Apr 19\, 2010 at 01&#8203;:34&#8203;:02AM \-0700\, perl@&#8203;0ne\.us wrote&#8203;:
> > I've been smoking CPAN modules for CPANTesters and noticed that my bots
> > got OOMKilled from time to time\.
> 
> Thanks for the report\.
> 
> It appears there are two \(mostly\) unrelated issues here\. First is the
> performance\, which is quadratic in string length\.
> 
> This is due to String&#8203;::Perl&#8203;::Warnings creating a regexp that looks
> essentially like /\.\*?\.\.\.\./\. This is equivalent to /^\.\*?\.\*?\.\.\.\.\./
> \(note the anchor\) and on repeated backtracking the two \.\*? make it go
> quadratic\. The fix for the String&#8203;::Perl&#8203;::Warnings module's author is to
> either remove the initial \.\*?\, or add an initial anchor\.
> 
> The second issue is the memory usage\, which again appears to be quadratic
> on string length\. This is due to the trie code leaking SVs on the tmps
> stack \(which then accumulate quadratically with the quadratic backtracking
> caused by the first issue\)\. I've currently got some ideas on how to rework
> that bit of code to not need to create temp SVs to store the accept
> states\, but in the meantime I've added a TODO test with commit
> 8bd05d90e59a554e79d73edf32bf35547dbc2c40

Now fixed in blead by the commit below\. Running the OPs code for 400
iterations on unpatched and patched blead gave the following results&#8203;:

&nbsp;    unpatched&#8203;: Loop 400&#8203;: 2\.76304984092712s rss\(214076 kB\)
&nbsp;    patched&#8203;:   Loop 400&#8203;: 2\.16697001457214s rss\(6976 kB\)

i\.e\. the memory leak is gone\, and it's a bit faster \(probably due to not
having to allocate lots of SVs in this pathological case\, rather than the
new trie code being intrinsically faster\)\.


commit 2e64971a6530d2645969bc489f564bfd3ce64993
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Mon May 3 13&#8203;:57&#8203;:58 2010 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Mon May 3 14&#8203;:29&#8203;:54 2010 \+0100

&nbsp;    tries&#8203;: don't allocate memory at runtime
&nbsp;    
&nbsp;    This is an indirect fix for
&nbsp;        \[perl \#74484\] Regex causing exponential runtime\+mem usage
&nbsp;    
&nbsp;    The trie runtime code was doing more SAVETMPS than FREETMPS and was thus
&nbsp;    growing a large tmps stack on heavy backtracking\. Rather than fixing this
&nbsp;    directly\, I rewrote part of the trie code so that it no longer needs to
&nbsp;    allocate memory in S\_regmatch \(it still does in find\_byclass\(\)\)\.
&nbsp;    
&nbsp;    The basic issue is that multiple branches in the trie may trigger an
&nbsp;    accept state; for example&#8203;:
&nbsp;    
&nbsp;        "abcd" =~ /xyz/abcd\.\*X|ab\.\*Y|/
&nbsp;    
&nbsp;    here\, words \(branches\) 2 and 3 are accept states\. The original approach
&nbsp;    was\, at run time\, to create a list of accepted word numbers and the
&nbsp;    character positions of the end of each of those words\. Then run the rest
&nbsp;    of the pattern for each word in the list in turn \(in word index order\)\.
&nbsp;    This requires memory for the list to be allocated and freed\.
&nbsp;    
&nbsp;    The new approach involves creating extra info at compile time; in
&nbsp;    particular\, for each word\, a pointer to the previous accepted word \(if
&nbsp;    any\) in the state tree\. For example for the above pattern\, part of the
&nbsp;    state tree may be
&nbsp;    
&nbsp;          q    b    c    d
&nbsp;        1 \-> 2 \-> 3 \-> 4 \-> 5
&nbsp;                \(\#3\)       \(\#2\)
&nbsp;    
&nbsp;    \(e\.g\. at state 1\, if the next char is 'a'\, we transition to state 2\)\.
&nbsp;    Here\, state 3 is an accept state with word \#3\, and 5 is an accept state
&nbsp;    with word \#2\. So we build a table indexed by word number\, which has
&nbsp;    wordinfo\[2\] = 3\, wordinfo\[3\] = 0\, thus building the word chain 2\->3\->0\.
&nbsp;    
&nbsp;    At run time we run the trie to completion\, and remember the word
&nbsp;    associated with the longest accept state \(word \#2 above\)\. Then by following
&nbsp;    back the chain of \.prev fields\, we can produce a list of all accepting
&nbsp;    words\. We then iteratively find the smallest\-numbered \(ie LH\-most\) word in
&nbsp;    the chain\, and run with it\. On failure and backtrack\, we find the
&nbsp;    next\-smallest and so on\.
&nbsp;    
&nbsp;    Since we are no longer recording the end\-position of each word in the
&nbsp;    string\, we have to recalculate this for each backtrack\. We initially
&nbsp;    record the end\-position of the shortest accepting word\, and given that we
&nbsp;    know the length of each word\, we can calculate the new position each time
&nbsp;    as an offset from that first word\. Depending on unicode and folding\, that
&nbsp;    calculation can be cheap or expensive\.
&nbsp;    
&nbsp;    This algorithm is optimised for the typical case where there are a small
&nbsp;    number \(\<= 2\) accepting states\.
&nbsp;    
&nbsp;    This patch creates a new compile\-time array\, trie\->wordinfo\[\]\, indexed by
&nbsp;    word number\, which contains relevant info about each word\. This also
&nbsp;    supersedes the old trie\->newword\[\] array\, whose function of recording
&nbsp;    "overspills" of multiple words per accept state\, is now handled as part of
&nbsp;    the wordinfo\[\]\.prev chain\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    ext/re/t/regop\.t
&nbsp;    M    regcomp\.c
&nbsp;    M    regcomp\.h
&nbsp;    M    regexec\.c
&nbsp;    M    regexp\.h
&nbsp;    M    t/op/svleak\.t


\-\- 
I don't want to achieve immortality through my work\.\.\. I want to achieve
it through not dying\.
&nbsp;    \-\- Woody Allen

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @demerphq

On 3 May 2010 15&#8203;:39\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> On Wed\, Apr 21\, 2010 at 12&#8203;:24&#8203;:21PM \+0100\, Dave Mitchell wrote&#8203;:
>> On Mon\, Apr 19\, 2010 at 01&#8203;:34&#8203;:02AM \-0700\, perl@&#8203;0ne\.us wrote&#8203;:
>> > I've been smoking CPAN modules for CPANTesters and noticed that my bots
>> > got OOMKilled from time to time\.
>>
>> Thanks for the report\.
>>
>> It appears there are two \(mostly\) unrelated issues here\. First is the
>> performance\, which is quadratic in string length\.
>>
>> This is due to String&#8203;::Perl&#8203;::Warnings creating a regexp that looks
>> essentially like /\.\*?\.\.\.\./\. This is equivalent to /^\.\*?\.\*?\.\.\.\.\./
>> \(note the anchor\) and on repeated backtracking the two \.\*? make it go
>> quadratic\. The fix for the String&#8203;::Perl&#8203;::Warnings module's author is to
>> either remove the initial \.\*?\, or add an initial anchor\.
>>
>> The second issue is the memory usage\, which again appears to be quadratic
>> on string length\. This is due to the trie code leaking SVs on the tmps
>> stack \(which then accumulate quadratically with the quadratic backtracking
>> caused by the first issue\)\. I've currently got some ideas on how to rework
>> that bit of code to not need to create temp SVs to store the accept
>> states\, but in the meantime I've added a TODO test with commit
>> 8bd05d90e59a554e79d73edf32bf35547dbc2c40
>
> Now fixed in blead by the commit below\. Running the OPs code for 400
> iterations on unpatched and patched blead gave the following results&#8203;:
>
>    unpatched&#8203;: Loop 400&#8203;: 2\.76304984092712s rss\(214076 kB\)
>    patched&#8203;:   Loop 400&#8203;: 2\.16697001457214s rss\(6976 kB\)
>
> i\.e\. the memory leak is gone\, and it's a bit faster \(probably due to not
> having to allocate lots of SVs in this pathological case\, rather than the
> new trie code being intrinsically faster\)\.
>
>
> commit 2e64971a6530d2645969bc489f564bfd3ce64993
> Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
> AuthorDate&#8203;: Mon May 3 13&#8203;:57&#8203;:58 2010 \+0100
> Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
> CommitDate&#8203;: Mon May 3 14&#8203;:29&#8203;:54 2010 \+0100
>
>    tries&#8203;: don't allocate memory at runtime
>
>    This is an indirect fix for
>        \[perl \#74484\] Regex causing exponential runtime\+mem usage
>
>    The trie runtime code was doing more SAVETMPS than FREETMPS and was thus
>    growing a large tmps stack on heavy backtracking\. Rather than fixing this
>    directly\, I rewrote part of the trie code so that it no longer needs to
>    allocate memory in S\_regmatch \(it still does in find\_byclass\(\)\)\.
>
>    The basic issue is that multiple branches in the trie may trigger an
>    accept state; for example&#8203;:
>
>        "abcd" =~ /xyz/abcd\.\*X|ab\.\*Y|/
>
>    here\, words \(branches\) 2 and 3 are accept states\. The original approach
>    was\, at run time\, to create a list of accepted word numbers and the
>    character positions of the end of each of those words\. Then run the rest
>    of the pattern for each word in the list in turn \(in word index order\)\.
>    This requires memory for the list to be allocated and freed\.
>
>    The new approach involves creating extra info at compile time; in
>    particular\, for each word\, a pointer to the previous accepted word \(if
>    any\) in the state tree\. For example for the above pattern\, part of the
>    state tree may be
>
>          q    b    c    d
>        1 \-> 2 \-> 3 \-> 4 \-> 5
>                \(\#3\)       \(\#2\)
>
>    \(e\.g\. at state 1\, if the next char is 'a'\, we transition to state 2\)\.
>    Here\, state 3 is an accept state with word \#3\, and 5 is an accept state
>    with word \#2\. So we build a table indexed by word number\, which has
>    wordinfo\[2\] = 3\, wordinfo\[3\] = 0\, thus building the word chain 2\->3\->0\.
>
>    At run time we run the trie to completion\, and remember the word
>    associated with the longest accept state \(word \#2 above\)\. Then by following
>    back the chain of \.prev fields\, we can produce a list of all accepting
>    words\. We then iteratively find the smallest\-numbered \(ie LH\-most\) word in
>    the chain\, and run with it\. On failure and backtrack\, we find the
>    next\-smallest and so on\.
>
>    Since we are no longer recording the end\-position of each word in the
>    string\, we have to recalculate this for each backtrack\. We initially
>    record the end\-position of the shortest accepting word\, and given that we
>    know the length of each word\, we can calculate the new position each time
>    as an offset from that first word\. Depending on unicode and folding\, that
>    calculation can be cheap or expensive\.
>
>    This algorithm is optimised for the typical case where there are a small
>    number \(\<= 2\) accepting states\.
>
>    This patch creates a new compile\-time array\, trie\->wordinfo\[\]\, indexed by
>    word number\, which contains relevant info about each word\. This also
>    supersedes the old trie\->newword\[\] array\, whose function of recording
>    "overspills" of multiple words per accept state\, is now handled as part of
>    the wordinfo\[\]\.prev chain\.

Hi Dave\. Thanks for this\. A very educational patch\.

Cheers\,
yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"
